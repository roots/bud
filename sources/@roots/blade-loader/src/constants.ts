export const scriptLoaders = [
  {loader: `@roots/blade-loader/script-loader`, type: `asset/source`},
  {loader: `@roots/blade-loader/asset-loader`, type: `asset/source`},
]

export const scriptPattern = /\.(js|mjs|jsx|ts|tsx|php)$/
