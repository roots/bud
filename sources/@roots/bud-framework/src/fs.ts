import type {Bud} from '@roots/bud-framework'
import type {Contract} from '@roots/bud-framework/service'

import {join} from 'node:path'

import {bind} from '@roots/bud-support/decorators/bind'
import {Filesystem, json, yml} from '@roots/bud-support/filesystem'
import globby from '@roots/bud-support/globby'
import isUndefined from '@roots/bud-support/lodash/isUndefined'
import logger from '@roots/bud-support/logger'
import {S3} from '@roots/filesystem'

/**
 * {@link Filesystem} service
 */
export class FS extends Filesystem implements Contract {
  /**
   * JSON
   *
   * @see {@link https://bud.js.org/docs/bud.fs/json}
   */
  public json: typeof json = json

  /**
   * {@link Contract.label}
   */
  public label = `fs`

  /**
   * S3
   *
   * @see {@link https://bud.js.org/docs/bud.fs/s3}
   */
  public s3?: S3

  /**
   * YML
   *
   * @see {@link https://bud.js.org/docs/bud.fs/yml}
   */
  public yml: typeof yml = yml

  /**
   * Class constructor
   */
  public constructor(public _app: () => Bud) {
    super(_app().context.basedir)
    this.s3 = new S3()
  }

  /**
   * {@link Contract.app}
   */
  public get app(): Bud {
    return this._app()
  }

  /**
   * {@link Contract.done}
   */
  public done() {
    return this.app
  }

  /**
   * {@link Contract.logger}
   */
  public get logger(): typeof logger {
    return logger.scope(`fs`)
  }

  /**
   * Set bucket
   *
   * @param bucket - {@link S3.bucket}
   *
   * @see {@link https://bud.js.org/docs/bud.fs/s3#setup}
   */
  @bind
  public setBucket(bucket: string) {
    this.app.after(async () => {
      this.s3.config.set(`bucket`, bucket)
    })

    return this
  }

  /**
   * Set credentials
   *
   * @param credentials - {@link S3.credentials}
   *
   * @see {@link https://bud.js.org/docs/bud.fs/s3#setup}
   */
  @bind
  public setCredentials(credentials: S3[`config`][`credentials`]) {
    this.app.after(async () => {
      this.s3.config.set(`credentials`, credentials)
    })

    return this
  }

  /**
   * Set endpoint
   *
   * @param endpoint - S3 endpoint
   *
   * @see {@link https://bud.js.org/docs/bud.fs/s3#setup}
   */
  @bind
  public setEndpoint(endpoint: S3[`config`][`endpoint`]) {
    this.app.after(async () => {
      this.s3.config.set(`endpoint`, endpoint)
    })

    return this
  }

  /**
   * Set S3 region
   *
   * @param region - S3 region
   *
   * @see {@link https://bud.js.org/docs/bud.fs/s3#setup}
   */
  @bind
  public setRegion(region: S3[`config`][`region`]) {
    this.app.after(async () => {
      this.s3.config.set(`region`, region)
    })

    return this
  }

  /**
   * Upload files to S3
   *
   * @param options - upload options
   *
   * @see {@link https://bud.js.org/docs/bud.fs/s3#uploading-files}
   */
  @bind
  public upload(options?: {
    destination?: string
    files?: string
    keep?: false | number
    source?: string
  }): this {
    const {destination, files, keep, source} = {
      destination: options?.destination,
      files: isUndefined(options?.files) ? `**/*` : options.files,
      keep: isUndefined(options?.keep) ? 5 : options.keep,
      source: isUndefined(options?.source)
        ? this.app.path(`@dist`)
        : options.source,
    }

    const s3Path = (path: string) =>
      destination ? join(destination, path) : path

    this.app.after(async () => {
      this.app.dashboard.updateStatus(`Deploying files to S3`)

      await globby(files, {cwd: source}).then(async files => {
        const descriptions = await Promise.all(
          files.map(async file => {
            const contents = await this.read(join(source, file), `buffer`)
            return {contents, file}
          }),
        )

        const manifestExists = await this.s3.exists(
          s3Path(`upload-manifest.json`),
        )

        const entries = Object.entries(
          manifestExists
            ? await this.s3
                .read(s3Path(`upload-manifest.json`))
                .then(this.json.parse)
            : {},
        )

        await Promise.all(
          descriptions.map(async ({contents, file}) => {
            this.logger.await(`Upload ${file} to ${this.s3.ident}`)

            try {
              await this.s3.write(s3Path(file), contents)
              this.logger.success(`Upload ${file} to ${this.s3.ident}`)
            } catch (error) {
              this.logger.error(`Upload ${file} to ${this.s3.ident}`)
              throw error
            }
          }),
        )

        if (typeof keep !== `number`) return

        const stale =
          keep > 1 ? entries.splice(0, entries.length - (keep - 1)) : []

        await Promise.all(
          [...new Set(stale)]
            .flatMap(([key, value]: [string, Array<string>]) => value)
            .filter(
              key =>
                !entries.some(([_, value]: [string, Array<string>]) =>
                  value.includes(key),
                ),
            )
            .map(async key => {
              const fileExists = await this.s3.exists(key)
              if (!fileExists) return

              this.logger.await(
                `Remove ${key} from ${this.s3.ident} (stale)`,
              )
              await this.s3.delete(key)
              this.logger.success(
                `Remove ${key} from ${this.s3.ident} (stale)`,
              )
            }),
        )

        this.logger.await(`Write upload-manifest.json to ${this.s3.ident}`)

        await this.s3.write({
          Body: Buffer.from(
            this.json.stringify({
              ...entries.reduce((acc, [k, v]) => ({...acc, [k]: v}), {}),
              [new Date().getTime()]: descriptions.map(({file}) => file),
            }),
          ),
          Key: s3Path(`upload-manifest.json`),
        })

        this.logger.success(
          `Write upload-manifest.json to ${this.s3.ident}`,
        )
      })
    })

    return this
  }
}
