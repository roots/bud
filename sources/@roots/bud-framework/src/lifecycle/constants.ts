/**
 * Service lifecycle events
 *
 * @public
 */
export const LIFECYCLE_EVENTS = [
  `bootstrap`,
  `bootstrapped`,
  `register`,
  `registered`,
  `boot`,
  `booted`,
]

/**
 * Services which are only instantiated in the parent compiler context.
 *
 * @public
 */
export const PARENT_SERVICES: Array<string> = [
  `@roots/bud-compiler`,
  `@roots/bud-dashboard`,
  `@roots/bud-server`,
]

/**
 * Services which are only instantiated in development
 *
 * @public
 */
export const DEVELOPMENT_SERVICES: Array<string> = [`@roots/bud-server`]
