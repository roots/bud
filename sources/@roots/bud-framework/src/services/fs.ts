import {join} from 'node:path'

import {Filesystem as FS, json, yml} from '@roots/bud-support/filesystem'
import globby from '@roots/bud-support/globby'
import isUndefined from '@roots/bud-support/lodash/isUndefined'
import {S3} from '@roots/filesystem/s3'

import type {Bud} from '../bud.js'

/**
 * Filesystem service
 */
export default class FileSystem extends FS {
  /**
   * Access {@link Bud}
   *
   * @public @readonly
   */
  public get app(): Bud {
    return this._app()
  }

  /**
   * Logger
   *
   * @public
   */
  public get logger() {
    const logger = this.app.context.logger.make(this.app.label, `fs`)

    return logger
  }

  /**
   * JSON handling
   *
   * @public
   */
  public json: typeof json = json

  /**
   * S3 instance
   *
   * @public
   */
  public s3: S3

  /**
   * YML handling
   *
   * @public
   */
  public yml: typeof yml = yml

  /**
   * Class constructor
   *
   * @public
   */
  public constructor(public _app: () => Bud) {
    super(_app().context.basedir)

    this.s3 = new S3()

    Object.assign(this.app, {
      json: this.json,
      yml: this.yml,
    })
  }

  /**
   * Set bucket
   *
   * @param bucket - {@link S3.bucket}
   * @public
   */
  public setBucket(bucket: string) {
    this.app.after(async (bud: Bud) => {
      bud.fs.s3.config.set(`bucket`, bucket)
    })

    return this
  }

  /**
   * Set credentials
   *
   * @param credentials - {@link S3.credentials}
   * @public
   */
  public setCredentials(credentials: S3[`config`][`credentials`]) {
    this.app.after(async (bud: Bud) => {
      bud.fs.s3.config.set(`credentials`, credentials)
    })

    return this
  }

  /**
   * Set endpoint
   *
   * @param options - upload options
   * @public
   */
  public setEndpoint(endpoint: S3[`config`][`endpoint`]) {
    this.app.after(async (bud: Bud) => {
      bud.fs.s3.config.set(`endpoint`, endpoint)
    })

    return this
  }

  /**
   * Set S3 region
   *
   * @param options - upload options
   * @public
   */
  public setRegion(region: S3[`config`][`region`]) {
    this.app.after(async (bud: Bud) => {
      bud.fs.s3.config.set(`region`, region)
    })

    return this
  }

  /**
   * Upload files to S3
   *
   * @param options - upload options
   * @public
   */
  public upload(options?: {
    source?: string
    destination?: string
    files?: string
    keep?: number | false
  }): this {
    const {source, files, keep, destination} = {
      source: isUndefined(options?.source)
        ? this.app.path(`@dist`)
        : options.source,
      files: isUndefined(options?.files) ? `**/*` : options.files,
      keep: isUndefined(options?.keep) ? 5 : options.keep,
      destination: options?.destination,
    }

    const s3Path = (path: string) =>
      destination ? join(destination, path) : path

    this.app.after(async (bud: Bud) => {
      await globby(files, {cwd: source}).then(async files => {
        const descriptions = await Promise.all(
          files.map(async file => {
            const contents = await bud.fs.read(
              join(source, file),
              `buffer`,
            )

            return {file, contents}
          }),
        )

        const manifestExists = await bud.fs.s3.exists(
          s3Path(`upload-manifest.json`),
        )
        const entries = Object.entries(
          manifestExists
            ? await bud.fs.s3
                .read(s3Path(`upload-manifest.json`))
                .then(bud.fs.json.parse)
            : {},
        )

        await Promise.all(
          descriptions.map(async ({file, contents}) => {
            bud.fs.logger.await(`Upload ${file} to ${bud.fs.s3.ident}`)

            try {
              await bud.fs.s3.write(s3Path(file), contents)
              bud.fs.logger.success(`Upload ${file} to ${bud.fs.s3.ident}`)
            } catch (error) {
              bud.fs.logger.error(`Upload ${file} to ${bud.fs.s3.ident}`)
              throw error
            }
          }),
        )

        if (typeof keep !== `number`) return

        const stale =
          keep > 1 ? entries.splice(0, entries.length - (keep - 1)) : []

        await Promise.all(
          [...new Set(stale)]
            .flatMap(([key, value]: [string, Array<string>]) => value)
            .filter(
              key =>
                !entries.some(([_, value]: [string, Array<string>]) =>
                  value.includes(key),
                ),
            )
            .map(async key => {
              const fileExists = await bud.fs.s3.exists(key)
              if (!fileExists) return

              bud.fs.logger.await(
                `Remove ${key} from ${bud.fs.s3.ident} (stale)`,
              )
              await bud.fs.s3.delete(key)
              bud.fs.logger.success(
                `Remove ${key} from ${bud.fs.s3.ident} (stale)`,
              )
            }),
        )

        bud.fs.logger.await(
          `Write upload-manifest.json to ${bud.fs.s3.ident}`,
        )

        await bud.fs.s3.write({
          Key: s3Path(`upload-manifest.json`),
          Body: Buffer.from(
            bud.fs.json.stringify({
              ...entries.reduce((acc, [k, v]) => ({...acc, [k]: v}), {}),
              [new Date().getTime()]: descriptions.map(({file}) => file),
            }),
          ),
        })

        bud.fs.logger.success(
          `Write upload-manifest.json to ${bud.fs.s3.ident}`,
        )
      })
    })

    return this
  }
}
