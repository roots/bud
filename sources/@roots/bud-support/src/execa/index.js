import {execa} from 'execa'
export default execa
