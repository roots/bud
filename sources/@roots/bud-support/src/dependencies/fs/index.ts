import fs from 'fs-extra'

export const {
  ensureDir,
  ensureFile,
  outputFile,
  pathExists,
  readFile,
  readJson,
  remove,
  writeFile,
  writeJson,
} = fs
