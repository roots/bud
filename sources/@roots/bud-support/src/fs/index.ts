import fs from 'fs-extra'

export const {
  copyFile,
  copy,
  ensureDir,
  ensureFile,
  outputFile,
  pathExists,
  readFile,
  readJson,
  remove,
  writeFile,
  writeJson,
} = fs
