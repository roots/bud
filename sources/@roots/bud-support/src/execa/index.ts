import {execa} from 'execa'

export default execa
export * from 'execa'
