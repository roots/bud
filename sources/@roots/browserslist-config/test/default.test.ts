import {path} from '@repo/constants'
import {execa} from 'execa'
import fs from 'fs-jetpack'
import {describe, expect, it} from 'vitest'

describe(`@roots/browserslist-config`, async () => {
  describe(`@roots/browserslist-config/default`, async () => {
    await execa(`yarn`, [`bud`, `build`], {
      cwd: path(
        `sources/@roots/browserslist-config/test/__fixtures__/default`,
      ),
    })

    it(`should compile as expected`, async () => {
      expect(
        await fs.readAsync(
          path(
            `sources/@roots/browserslist-config/test/__fixtures__/default/dist/css/main.css`,
          ),
        ),
      ).toMatchSnapshot()
    })
  })

  describe(`@roots/browserslist-config/last-2-versions`, async () => {
    await execa(`yarn`, [`bud`, `build`], {
      cwd: path(
        `sources/@roots/browserslist-config/test/__fixtures__/last-2-versions`,
      ),
    })

    it(`should compile as expected`, async () => {
      expect(
        await fs.readAsync(
          path(
            `sources/@roots/browserslist-config/test/__fixtures__/last-2-versions/dist/css/main.css`,
          ),
        ),
      ).toMatchSnapshot()
    })
  })

  describe(`@roots/browserslist-config/last-3-versions`, async () => {
    await execa(`yarn`, [`bud`, `build`], {
      cwd: path(
        `sources/@roots/browserslist-config/test/__fixtures__/last-3-versions`,
      ),
    })

    it(`should compile as expected`, async () => {
      expect(
        await fs.readAsync(
          path(
            `sources/@roots/browserslist-config/test/__fixtures__/last-3-versions/dist/css/main.css`,
          ),
        ),
      ).toMatchSnapshot()
    })
  })

  describe(`@roots/browserslist-config/wordpress`, async () => {
    await execa(`yarn`, [`bud`, `build`], {
      cwd: path(
        `sources/@roots/browserslist-config/test/__fixtures__/wordpress`,
      ),
    })

    it(`should compile as expected`, async () => {
      expect(
        await fs.readAsync(
          path(
            `sources/@roots/browserslist-config/test/__fixtures__/wordpress/dist/css/main.css`,
          ),
        ),
      ).toMatchSnapshot()
    })
  })
})
