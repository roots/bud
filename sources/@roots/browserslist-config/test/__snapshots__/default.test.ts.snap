// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@roots/browserslist-config > @roots/browserslist-config/default > should compile as expected 1`] = `".foo{accent-color:auto;bottom:0;left:0;margin-left:1rem;right:0;text-wrap:balance;top:0;word-break:auto-phrase}"`;

exports[`@roots/browserslist-config > @roots/browserslist-config/last-2-versions > should compile as expected 1`] = `".foo{accent-color:auto;bottom:0;left:0;margin-left:1rem;right:0;text-wrap:balance;top:0;word-break:auto-phrase}"`;

exports[`@roots/browserslist-config > @roots/browserslist-config/last-3-versions > should compile as expected 1`] = `".foo{accent-color:auto;bottom:0;left:0;margin-left:1rem;right:0;text-wrap:balance;top:0;word-break:auto-phrase}"`;

exports[`@roots/browserslist-config > @roots/browserslist-config/wordpress > should compile as expected 1`] = `".foo{accent-color:auto;inset:0;margin-inline-start:1rem;text-wrap:balance;word-break:auto-phrase}"`;
